### ISSUE TITLE
[Bug] There is no limit of how many top-sites you can add

### ISSUE THREAD

**Issue by abodea** [mozilla-mobile/fenix#8312]
**Update from UX:**

- To keep the design scalable, let's limit top sites to 16 (2 x 8 swipe-able, see screenshot and please ignore the pin icon on the first three tabs for now)
- Once people add their 9th top site we show dots below top sites to indicate that there's  more 
- People can swipe (similar to how it works in Fennec) to see their 9th-16th top site
- When people are about to add a 17th top site a dialog is displayed telling them that they reached the max amount for top sites 

![image](https://user-images.githubusercontent.com/353960/84441360-edbf3300-ac3b-11ea-87fc-98f7fa40b7ec.png)

-------
## Steps to reproduce
1. Launch Fenix.
2. Add 50+ top-sites.

### Expected behavior
There should be a limit regarding how many top-sites you can add.
### Actual behavior
50+ top-sites can be added.
### Device information

* Android device: **Samsung Galaxy S10+(Android 9)**
* Fenix version: **Latest Nightly from 2/11.**

![20200211-175517](https://user-images.githubusercontent.com/42831109/74253850-4d421c00-4cf8-11ea-9ba8-99a0af942d9c.gif)

**Comment by brampitoyo** [comment]
Paging @topotropic for feedback. Should there be a limit to how many items users can add under Top Sites?

@mcarare just to be aware of the fact that most users will probably not create enough Top Sites to ever see this problem.

**Comment by AmyYLee** [comment]
@topotropic to follow-up

**Comment by topotropic** [comment]
Let's limit to 8 top sites (2 rows a 4 sites) – when people add the 9th top site we remove the oldest.

**Comment by cadeyrn** [comment]
@topotropic:

I already saw a screenshot of a user with more top sites. Why should this be limited at all? Sure, it doesn't look great but it's the user's decision, no? I don't need top sites at all but if it's useful for someone to have 20 top sites or even more why should this not be allowed? Is there a technical reason why it should be limited? Unfortunately there is no explanation in this issue what the benefit of such a limit is.

Please also note that with the new tabs tray and the removal of the tabs from the start screen there is more room that could be used for top sites now - when the user wants it.

Also I am not convinved that it's a good idea to remove the oldest top site when trying to add more top sites as allowed. If I were to use the top sites feature, I would probably add the most important website at first. But with this proposal the website I consider as the most important will be the first website that get removed again.

**Comment by liuche** [comment]
We want to keep our design scalable for other things we want to add to the homescreen (synced tabs, bookmarks, etc)

Should be a non-destructive method.

**Comment by liuche** [comment]
for release minimum, we need to have a way to cap this before release, but can improve it in subsequent releases.

**Comment by betsymi** [comment]
Strings for the dialog when top site limit is reached. 

**Top site limit reached**

To add a new top site, remove one. Long press the site and select remove. 

OK, Got It

**Comment by AmyYLee** [comment]
As discussed with @gabrielluong we are not setting a limit on how many topsites you can add since we cannot accommodate  our various screen layouts and sizes (tablet and landscape) with a set number of top sites.

**Comment by ShatteredPixel** [comment]
Can somebody here please clarify me that you are not going to limit the number of Top Sites I can add. Top Sites are a very important feature for me. And they act like a Speed Dial for me, I use them very frequently regularly. And I'm not with having a swipe menu either,  there's reason why I called them Speed Dials, they should be accessible very speedily. As said in the comment here: https://github.com/mozilla-mobile/fenix/issues/8312#issuecomment-607006900 most users are never going to have too many top sites to have the issue this issue is mentioning. I'm a very avid user of Top Sites and I have never created too many top sites. My current top sites are about 30.

@topotropic please don't remove the last oldest top sites or any of them! Top Sites are an extremely important feature for me and I cannot afford to lose a single Top site, in fact I'm going to file a new issue asking to add an undo button when each top site is deleted. 

Please keep this features behaviour as it is right now there's no need to add or remove anything to related to number of Top Sites, nobody is going to create to many top sites and even if they do it it should be there choice they will do this by keeping in mind that other parts of the homescreen are going to be hidden and they will know this. Please, I request you to don't limit the number of sites I can add to top sites or try to delete them suddenly out of nowhere! 
Please...


Thanks!

**Comment by cadeyrn** [comment]
@gabrielluong I really like the idea of re-using the concept of Firefox Lite to have multiple slides for users with more top sites. It solves the problem of keeping the design scalable but it still allows users to add more than eight top sites.

With this concept in place is it really necessary to limit the top sites to a maximum of two slides for users who really want to use more top sites? I agree it's useful to have a limit because 100 slides (to be extreme) would obviously cause other issues. But it should be possible without a big risk to give the user the option to have… let's say… five rows. What do you think?

**Comment by gabrielluong** [comment]
From @topotropic on the pager dot colours,

```
lighter color is <color name=“light_grey_30”>#E0E0E6</color>
and darker color <color name=“light_grey_50">#BFBFC9</color>
```

**Comment by sheikh-azharuddin** [comment]
The new top site design is bad...💔
The screen can definitely accommodate 5 icons so why 4 icons?  Also why just 2 rows for top sites when there is too much space in the homescreen?? 

![Screenshot_20200825-134531641](https://user-images.githubusercontent.com/58127841/91149962-5e250e80-e6d9-11ea-8230-056d298aadb5.jpg)

**Comment by sheikh-azharuddin** [comment]
there must be a customisable option to configure number of columns and number of rows like android quick toggle panel

**Comment by cadeyrn** [comment]
>  The screen can definitely accommodate 5 icons so why 4 icons? Also why just 2 rows for top sites when there is too much space in the homescreen??

This was already answered:

> We want to keep our design scalable for other things we want to add to the homescreen (synced tabs, bookmarks, etc)

Also note that for other users collections are shown on the homescreen.

**Comment by sheikh-azharuddin** [comment]
Uff this synced tabs and collection.... These are promoted everywhere... The desktop version has synced tabs setting hidden but in mobile version this is being promoted too much!
The collection is being given too much importance compared to top sites when there is literally no difference between these 2... Unless collections are synced...these add an extra step to access as compared to top sites. 

After reading this many will see sync feature is being considered as per #xxxxx number which is literally I am hearing since months!
The homepage should be customisable as per as user choice...in particularly the number of number of icons and rows should be flexible so that user can change as per as device screen size or need . I have a large screen device and icons now feels too far apart. This is probably because the icons have fix size px rather than scable size%

In a nutshell now it is difficult to access top sites as I need to take one more step to scroll.. Not to mention there is no option to rearrange the top sites adding another frustration!

**Comment by cadeyrn** [comment]
> [collections] These are promoted everywhere

For the startscreen already solved by #12377. It doesn't belong to this issue.

> Unless collections are synced

That's #6512. It doesn't belong to this issue.

> The homepage should be customisable as per as user choice

This depends on the specific features on the startscreen. So far everything can be removed from the home screen. I can't speak about future features that are not yet implemented. It doesn't belong to this issue.

> in particularly the number of number of icons and rows should be flexible so that user can change as per as device screen size or need

It doesn't make much sense to customize the number of items per row. It has to work out of the box. But as I already told you - that's #13765 and doesn't belong to this issue.

The number of rows doesn't need to be customizable with the new slides implementation because it solves the same problem. And that more than only two slides should be allowed it's exactly what I proposed in this issue.

> Not to mention there is no option to rearrange the top sites adding another frustration!

That's #8305. It doesn't belong to this issue.

**Comment by madb1lly** [comment]
Hi all,

I don't see the benefit in having a limit. In any case by only allowing 8 on screen at a time the limit really becomes 8, regardless of how many slides there are.

Any limit could be compensated for by finally adding #10342 (frequently used sites) which is a feature I'm sure many people migrated from Fennec will miss.  

Cheers 🙂

**Comment by cadeyrn** [comment]
> In any case by only allowing 8 on screen at a time the limit really becomes 8, regardless of how many slides there are.

Nope. Did you already try it? I already know this from Firefox Lite and it's really not that bad. It's definitively not the same as only having one slide with a limit.

> Any limit could be compensated for by finally adding #10342 (frequently used sites) which is a feature I'm sure many people migrated from Fennec will miss.

I don't understand this part. Auto populated sites won't change the limit so it shouldn't change anyhing for this issue.  Could you explain what you mean?

**Comment by sheikh-azharuddin** [comment]
But why have slides when whole screen can be utilized... You say screen can be used for other things like synced tabs and bookmarks.... Do you really think the screen can accommodate all the bookmarks? It is just a mini screen device 5-6" and everyone is expectating to fit everything! The synced tab is being promoted everywhere... Tab menu, 3 dot menu and the only place left was home screen... Now i am hearing there also  it will be added😓

@cadeyrn for some things you said this is referenced here and there and not related to this which I don't understand... When you do a change it directly or indirectly affect other things and user experience... The number of icons and layout is limited so obviously user will demand customizations but you say that one is part of another request! 😬😬

So fix those issues first and then limit the features....

**Comment by cadeyrn** [comment]
> You say screen can be used for other things like synced tabs and bookmarks.... Do you really think the screen can accommodate all the bookmarks?

I didn't say this, it was said by a Mozilla team member and I quoted it. I can't answer your question because I can' talk about a feature that is not yet implemented. But so far nobody said that _all_ bookmarks will be displayed on the homescreen. But again: It's the wrong issue to discuss this.

> The synced tab is being promoted everywhere... Tab menu, 3 dot menu and the only place left was home screen... Now i am hearing there also it will be added😓

It was mentioned as an example, I don't think it's already clear at all how the new homescreen will look like. There is still no UX spec and we have to wait to see what Mozilla's plans are. But again: It's the wrong issue to discuss this.

> @cadeyrn for some things you said this is referenced here and there and not related to this which I don't understand... When you do a change it directly or indirectly affect other things and user experience... The number of icons and layout is limited so obviously user will demand customizations but you say that one is part of another request! 😬😬

Please just read the title of the issue: This issue is about the limit of how many top sites you can add. It's not about the layout of the top sites, not about collections, not about synced tabs, not about bookmarks. This is no meta issue about every aspect of the homescreen. It's about one specific task and I told you the issues numbers where you can discuss the other topics.

**Comment by madb1lly** [comment]
Hi @sheikh-azharuddin,

Here's a signpost I hope you find helpful => #12065.

Cheers 🙂

**Comment by sheikh-azharuddin** [comment]
> Hi @sheikh-azharuddin,
> 
> Here's a signpost I hope you find helpful => #12065.
> 
> Cheers 🙂

Developer own story🤣🤣
I am using Firefox till it is user friendly and secure ....i think that day is near when I will switch to edge or chrome 🙏🏼

**Comment by yoasif** [comment]
Displeasure in the wild: https://www.reddit.com/r/firefox/comments/igheq5/current_firefox_version_made_me_remove_it_from/g2u00sr/

>even worse today nighlty build had redisigned home page we can have 8 top sites at max, this is unbearable I finally switched to a Samsung internet and as for now zero regrets

Agree with some other folks here - I don't quite understand why there is a limit at all. Pagination makes it scalable to begin with.

**Comment by s-ankur** [comment]
Hi, this looks really weird when you set the android display size to smallest(from android settings)
![Screenshot_20200826-020302_Firefox_Nightly](https://user-images.githubusercontent.com/33252312/91224905-65333780-e740-11ea-9d7f-779dc181023b.png)
I don't see why we need to have the hard limit at 8 when the screen can clearly accomodate more in the same space. 
As a side note, it would be great if all ui changes were tested with different display sizes

**Comment by cadeyrn** [comment]
> this looks really weird

https://github.com/mozilla-mobile/fenix/pull/14116#issuecomment-679257809:

> The proper layout of the items and grid will be done in #13077

**Comment by s-ankur** [comment]
@cadeyrn I am not sure how the layout could be fixed while still limiting the max number of top sites to 8(which I think is the current issue). Forgive me if I am missing something obvious here

**Comment by cadeyrn** [comment]
> (I think you meant to quote #13765.)

Thanks. I have no idea how i managed to change the number when copying the quote. I fixed my previous comment.

**Comment by madb1lly** [comment]
Hi all,

Looking at what we now have, there is no way this issue should have been marked bug - it's a usability preference at best not a bug. Personally I consider the new top sites layout more of a bug than the previous one.

Cheers 🙂

**Comment by sheikh-azharuddin** [comment]
> Hi all,
> 
> Looking at what we now have, there is no way this issue should have been marked bug - it's a usability preference at best not a bug. Personally I consider the new top sites layout more of a bug than the previous one.
> 
> Cheers 🙂

The new layout is simply trash! Also check user frustration in reddit

**Comment by cadeyrn** [comment]
> is simply trash!

https://www.mozilla.org/en-US/about/governance/policies/participation/:

> Be Respectful
>
> Value each other’s ideas, styles and viewpoints. We may not always agree, but disagreement is no excuse for poor manners. Be open to different possibilities and to being wrong. Be respectful in all interactions and communications, especially when debating the merits of different options.

**Comment by sheikh-azharuddin** [comment]
Uff sorry @cadeyrn Mozilla fanboy.... Do Mozilla listens..or they simply do what they think is right! .

**Comment by klint** [comment]
@cadeyrn : you said that the current limit is 2 slides? But I had far more than 16 top sites before this was implemented and now I have slides which I can swipe to. The only issue is that there are only 2  progress dots instead of 5.

**Comment by klint** [comment]
Another comment on the swipe action itself : maybe because it is an horizontal swipe on a narrow band on the top of the screen, I find it rather uneasy to swipe. It often ends up in a diagonal gesture that actually scrolls the page down instead of scrolling through the slides...

**Comment by cadeyrn** [comment]
> Uff sorry @cadeyrn Mozilla fanboy....

I am a "fanboy" because I asked you not to be disrepectful. Wow, against such a logic it's hard to argue… Really, it's not okay how you interact with other **people**! 😐

>  you said that the current limit is 2 slides? But I had far more than 16 top sites before this was implemented and now I have 5 slides which I can swipe to.

Interesting. Maybe I misunderstood the code? @gabrielluong should be able to clarify. 😊

**Comment by gabrielluong** [comment]
Currently, it is indeed designed to be a max of 16 top sites displayed in 4 x 2 grid. We can revisit the max of 16 top sites. I imagine the rationale for the 4 x 2 grid and paging is to preserve the vertical space of the home screen. This is an active development space, so expect continual improvements.

**Comment by klint** [comment]
> > you said that the current limit is 2 slides? But I had far more than 16 top sites before this was implemented and now I have 5 slides which I can swipe to.
> 
> Interesting. Maybe I misunderstood the code? @gabrielluong should be able to clarify. 😊
@cadeyrn It seems the code was changed yesterday and I'm able to swipe through 2 slides only, now. Which I find quite sad, as I don't really understand why the slides should be limited to 2 (and 16 top sites), after being able to play with more yesterday :)

**Comment by klint** [comment]
> 
> 
> Another comment on the swipe action itself : maybe because it is an horizontal swipe on a narrow band at the top of the screen, I find it rather uneasy to swipe. It often ends up in a diagonal gesture that actually scrolls the page down instead of scrolling through the slides...

FYI, I have opened a new issue #14287 that is somehow related to my remark above

**Comment by tortino** [comment]
I don't agree on limiting top sites: I use many and don't want any limit, it's my decision how much top sites I should see.
However I agree about adding "dots below top sites to indicate that there's more", once reached a limit (maybe 2 rows?).
But ONLY if fenix remembers the user choice about showing all top sites or showing less rows and the dots. 
If the user want to see all top sites he just taps on the dots, and fenix will always show all top sites, without limits.

**Comment by OUTDOOOR** [comment]
User customisation of rows and columns could be very useful. Current solution with two sides of the Top Pages is not comfortable and is much worse than previous version. Sliding between two sides is really annoying. I hope, that in next updates it will be improved.

**Comment by LaurentiuApahideanSV** [comment]
I have tested the issue on Firefox Preview Nightly 200831 (Build #2015761139). Only 16 websites can be added to topsites, however when trying to add the 17th top site there is no dialog informing the user that they reached the max amount of top sites.
### Devices used:
- OnePlus 6T (Android 9)
- Huawei MediaPad M3 (Android 7.0)

**Comment by gabrielluong** [comment]
> I have tested the issue on Firefox Preview Nightly 200831 (Build #2015761139). Only 16 websites can be added to topsites, however when trying to add the 17th top site there is no dialog informing the user that they reached the max amount of top sites.
> 
> ### Devices used:
> * OnePlus 6T (Android 9)
> * Huawei MediaPad M3 (Android 7.0)

@LaurentiuApahideanSV Thanks for the catch! I have filed #14529 for the dialog. This wasn't a release blocker for this feature so I didn't end up implementing it just yet.

**Comment by gabrielluong** [comment]
@LaurentiuApahideanSV given the above, should we say this is qa-verified?

**Comment by sv-ohorvath** [comment]
@gabrielluong I just tried this and without that limitation, users can re-try adding a top site 10 times until they give up. Doing so, will create a queue of 10 identical items to add to the top sites list, and once you remove a top site it gets added. If you have the same website 10x times in that list, you will see it and have to remove it 10 times to get rid of it.

**Comment by LaurentiuApahideanSV** [comment]
@gabrielluong Sites present on the Top Sites section do have an option to remove them but two issue have been encountered regarding  this: https://github.com/mozilla-mobile/fenix/issues/14538 and https://github.com/mozilla-mobile/fenix/issues/11188. I will mark this issue as verfied as all other encountered issues have been reported.

**Comment by madb1lly** [comment]
Hi all,

The new larger icons introduced today (tonight? 😉) in Nightly are much better, #13765. However, On my Nokia 8 it's still not a good use of space, I could comfortably fit 6 icons per row.

Maybe if the spacing between icons in rows and columns was the same it would be more aesthetically pleasing too, because it would be a grid of squares rather than a grid of rectangles.

Cheers 🙂

**Comment by gabrielluong** [comment]
Closing this issue now that it is eng:qa:verified

**Comment by LaurentiuApahideanSV** [comment]
Verified as fixed on Firefox Preview Beta 81.1.0-beta.2 (Build #2015761657).

### Devices used:
- Huawei MediaPad M3 (Android 7.0)
- OnePlus 6T (Android 9)
